File: coco_scala/src/main/scala/exp/m0/M0.scala
package exp.m0

trait Exp[T] extends exp.Exp[T] with Factory[T] {
  def eval: Double
}

trait Factory[T] extends exp.Factory[T] {
  def lit(value:Double) : exp.Exp[T]
  def add(left:exp.Exp[T], right:exp.Exp[T]) : exp.Exp[T]

  implicit override def convert(toConvert:exp.Exp[T]) : Exp[T]
}

trait Lit[T] extends Exp[T] {
  def value: Double

  def eval: Double = value
}

trait Add[T] extends Exp[T] {
  def left : exp.Exp[T]
  def right : exp.Exp[T]

  def eval: Double = left.eval + right.eval
}

object finalized {
  trait Exp extends exp.m0.Exp[Exp] with Factory {
    def getSelf : Exp = this
  }

  trait Factory extends exp.m0.Factory[Exp] {
    override def lit(value: Double) : Exp = new Lit(value)
    override def add(left: exp.Exp[Exp], right: exp.Exp[Exp]) : Exp = new Add(left, right)

    override implicit def convert(toConvert: exp.Exp[Exp]): Exp = toConvert.getSelf
  }

  class Lit(val value:Double) extends Exp with exp.m0.Lit[Exp]
  class Add(val left:exp.Exp[Exp], val right:exp.Exp[Exp]) extends Exp with exp.m0.Add[Exp]
}

File: coco_csharp/CoCo.Tests/M0.cs
using System;
using Xunit;
using EP.M0;

namespace CoCo.Tests.M0
{
    public interface TestTemplate<FT> : Factory<FT> {
        public void test()
        {
            Exp<FT> expr1 = convert(add(lit(1.0), lit(2.0)));
            Assert.Equal(3.0, expr1.eval());

            Exp<FT> expr2 = convert(lit(2.0));
            Assert.Equal(2.0, expr2.eval());

            Assert.Equal(3.0, convert(add(lit(1.0), lit(2.0))).eval());
            Assert.Equal(5.0, convert(lit(5.0)).eval());
        }
    }

    public class ActualTest : TestTemplate<EP.M0.Finalized.Exp>, EP.M0.Finalized.Factory {}

    public class M0Test
    {
        [Fact]
        public void TestM0()
        {
            TestTemplate<EP.M0.Finalized.Exp> theTest = new ActualTest();
            theTest.test();
        }
    }
}
